### PR TITLE
CLI command to mark table/schema for skipping in the table migration process.

### LINKED ISSUE #672: Add `databricks labs ucx skip --schema ... --table ...` command
`databricks labs ucx skip --schema ... --table ...`  to skip migration of a schema or a table and persist that into DBPROPERTIES and/or TBLPROPERTIES

### PR DESCRIPTION
**Pull request by FastLee**
closes #672

### CODE REVIEW

File: src/databricks/labs/ucx/cli.py [hunk 1]
```diff
@@ -34,10 +36,51 @@ def list_installations():
...
+    if not schema:
+        logger.error('"schema" is a required parameter.')
+        return None
+    ws = WorkspaceClient()
+    installation_manager = WorkspaceInstaller(ws)
+    logger.info("Fetching installation config.")
+    try:
+        sql_backend = StatementExecutionBackend(ws, installation_manager._current_config.warehouse_id)
+    except NotFound:
+        logger.error(
```
Comment (dmoore247) [inline src/databricks/labs/ucx/cli.py:52]: Need unit tests to cover

File: src/databricks/labs/ucx/cli.py [hunk 2]
```diff
@@ -34,10 +36,51 @@ def list_installations():
...
+            "Couldn't find UCX configuration in the user's home folder. "
+            "Make sure the current user has configured and installed UCX."
+        )
+        return None
+
+    if table:
+        try:
+            table_properties = sql_backend.fetch(f"SHOW TBLPROPERTIES hive_metastore.`{schema}`.`{table}`")
```
Comment (dmoore247) [inline src/databricks/labs/ucx/cli.py:60]: Elsewhere in the code, ws.table.get().properties and ws.table.list...table.properties are used. Please reconcile and standardize table property method fetch

File: src/databricks/labs/ucx/cli.py [hunk 3]
```diff
@@ -34,10 +36,51 @@ def list_installations():
...
+        except BadRequest:
+            logger.error(f"{schema}.{table} wasn't found in this workspace")
```
Comment (dmoore247) [inline src/databricks/labs/ucx/cli.py:62]: Log error details incase some other error generates `BadRequest`

Comment (dmoore247) [reply src/databricks/labs/ucx/cli.py:62]: Also, is it expected to not find tables in a workspace? If so then it's not really an error.
If a table reaches this point and is not found then these failures should be counted.

File: src/databricks/labs/ucx/cli.py [hunk 4]
```diff
@@ -34,10 +36,51 @@ def list_installations():
...
+            return None
+        if {_.key for _ in table_properties if _.key == "upgraded_to"}:
+            logger.error("Table was already migrated to UC")
```
Comment (dmoore247) [inline src/databricks/labs/ucx/cli.py:65]: This isn't really an error, a warning or info.
The message should include the table name.

File: src/databricks/labs/ucx/cli.py [hunk 5]
```diff
@@ -34,10 +36,51 @@ def list_installations():
...
+        logger.info(f"Setting up table {schema}.{table} to be skipped.")
+        sql_backend.execute(f"ALTER TABLE `{schema}`.`{table}` SET TBLPROPERTIES('UCX_SKIP' = true)")
+    else:
+        try:
+            sql_backend.fetch(f"DESCRIBE hive_metastore.`{schema}`")
+        except BadRequest:
+            logger.error(f"{schema} wasn't found in this workspace")
+            return None
+        logger.info(f"Setting up schema {schema} to be skipped.")
+        sql_backend.execute(f"ALTER SCHEMA `{schema}` SET DBPROPERTIES('UCX_SKIP' = true)")
```
Comment (dmoore247) [inline src/databricks/labs/ucx/cli.py:76]: Need unit tests (and integration tests)

**Review by dmoore247 — CHANGES_REQUESTED** [review]
Logging and testing issues.

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/cli.py [hunk 6]
```diff
@@ -34,10 +36,51 @@ def list_installations():
...
+    schema = kwargs["schema"]
+    table = kwargs["table"]
+    if not schema:
+        logger.error('"schema" is a required parameter.')
+        return None
+    ws = WorkspaceClient()
+    installation_manager = WorkspaceInstaller(ws)
+    logger.info("Fetching installation config.")
+    try:
+        sql_backend = StatementExecutionBackend(ws, installation_manager._current_config.warehouse_id)
```
[the remark this answers is not in the document]
Comment (FastLee) [reply src/databricks/labs/ucx/cli.py]: Done

File: src/databricks/labs/ucx/cli.py [hunk 7]
```diff
@@ -34,10 +37,36 @@ def list_installations():
     print(json.dumps(all_users))
 
 
+def skip(schema: str, table: str | None = None):
+    logger.info("Running skip command")
+    if not schema:
+        logger.error("--Schema is a required parameter.")
+        return None
+    ws = WorkspaceClient()
```
Comment (dmoore247) [inline src/databricks/labs/ucx/cli.py:45]: How does this work with `--profile` ?

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 8]
```diff
@@ -245,3 +246,37 @@ def _init_seen_tables(self):
...
+        # Marks a table to be skipped in the migration process
+        try:
+            self._backend.execute(f"ALTER TABLE `{schema}`.`{table}` SET TBLPROPERTIES('UCX_SKIP' = true)")
+        except NotFound as nf:
+            if "[TABLE_OR_VIEW_NOT_FOUND]" in str(nf):
+                logger.error(f"Failed to apply skip marker for Table {schema}.{table}. Table not found.")
+            else:
+                logger.error(nf)
+        except BadRequest as br:
+            logger.error(br)
```
Comment (dmoore247) [inline src/databricks/labs/ucx/hive_metastore/tables.py:270]: Please add unit & integration tests

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 9]
```diff
@@ -245,3 +246,37 @@ def _init_seen_tables(self):
...
+
+    def skip_schema(self, schema: str):
+        # Marks a schema to be skipped in the migration process
+        try:
+            self._backend.execute(f"ALTER SCHEMA `{schema}` SET DBPROPERTIES('UCX_SKIP' = true)")
```
Comment (dmoore247) [inline src/databricks/labs/ucx/hive_metastore/tables.py:275]: Properties generally follow the dot notation convention.
More extendable property name would be `databricks.labs.ucx.skip = true`

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 10]
```diff
@@ -245,3 +246,37 @@ def _init_seen_tables(self):
...
+        except NotFound as nf:
+            if "[SCHEMA_NOT_FOUND]" in str(nf):
+                logger.error(f"Failed to apply skip marker for Schema {schema}. Schema not found.")
+            else:
+                logger.error(nf)
+        except BadRequest as br:
+            logger.error(br)
```
Comment (dmoore247) [inline src/databricks/labs/ucx/hive_metastore/tables.py:282]: Please add unit tests

**Review by dmoore247 — CHANGES_REQUESTED** [review]
Let's have a more scalable property naming convention using the dot notation.

File: src/databricks/labs/ucx/cli.py [hunk 11]
```diff
@@ -34,10 +37,36 @@ def list_installations():
...
+    installation_manager = WorkspaceInstaller(ws)
+    logger.info("Fetching installation config.")
+    try:
+        warehouse_id = installation_manager._current_config.warehouse_id
+        sql_backend = StatementExecutionBackend(ws, warehouse_id)
```
Comment (nfx) [inline src/databricks/labs/ucx/cli.py:50]: look at https://github.com/databrickslabs/ucx/commit/65bffc55bd26fc1958cdfd2804a87c0e6c1d98ab to see what are the expectations for testing a command.

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 12]
```diff
@@ -245,3 +246,37 @@ def _init_seen_tables(self):
 
     def _table_already_upgraded(self, target) -> bool:
         return target in self._seen_tables
+
+
+class TablesAnnotator:
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/tables.py:251]: can you move it to `mapping.TablesMapping`? see https://github.com/databrickslabs/ucx/commit/65bffc55bd26fc1958cdfd2804a87c0e6c1d98ab#diff-231c20000db14946278e9cf9acf5e7ff20f91a527156def658929d962a9e754c

File: tests/unit/hive_metastore/test_tables.py [hunk 13]
```diff
@@ -107,3 +112,29 @@ def test_tables_returning_error_when_describing():
     tc = TablesCrawler(backend, "default")
     results = tc._crawl()
     assert len(results) == 1
+
+
+def test_skip_happy_path(mocker, caplog):
+    ws = mocker.patch("databricks.sdk.WorkspaceClient.__init__")
+    sbe = mocker.patch("databricks.labs.ucx.framework.crawlers.StatementExecutionBackend.__init__")
```
Comment (nfx) [inline tests/unit/hive_metastore/test_tables.py:119]: use `MockBackend` to verify the right queries are issued.

File: tests/unit/hive_metastore/test_tables.py [hunk 14]
```diff
@@ -107,3 +112,29 @@ def test_tables_returning_error_when_describing():
...
+    annotate = TablesAnnotator(ws, sbe)
+    annotate.skip_table(schema="schema", table="table")
+    assert len(caplog.records) == 0
+
+
+def test_skip_missing_schema(mocker, caplog):
+    ws = mocker.patch("databricks.sdk.WorkspaceClient.__init__")
+    sbe = mocker.patch("databricks.labs.ucx.framework.crawlers.StatementExecutionBackend.__init__")
```
Comment (nfx) [inline tests/unit/hive_metastore/test_tables.py:127]: use `MockBackend` to verify the right queries are issued.

File: tests/unit/hive_metastore/test_tables.py [hunk 15]
```diff
@@ -107,3 +112,29 @@ def test_tables_returning_error_when_describing():
...
+    sbe.execute.side_effect = NotFound("[SCHEMA_NOT_FOUND]")
+    annotate = TablesAnnotator(ws, sbe)
+    annotate.skip_schema(schema="schema")
+    assert [rec.message for rec in caplog.records if "schema not found" in rec.message.lower()]
+
+
+def test_skip_missing_table(mocker, caplog):
+    ws = mocker.patch("databricks.sdk.WorkspaceClient.__init__")
+    sbe = mocker.patch("databricks.labs.ucx.framework.crawlers.StatementExecutionBackend.__init__")
```
Comment (nfx) [inline tests/unit/hive_metastore/test_tables.py:136]: use `MockBackend` to verify the right queries are issued.

File: tests/unit/test_cli.py [hunk 16]
```diff
@@ -0,0 +1,34 @@
+import pytest
+from databricks.sdk.errors import NotFound
+from databricks.sdk.service import iam
+from databricks.sdk.service.iam import User
+
+from databricks.labs.ucx.cli import skip
+
+
+@pytest.fixture
+def ws(mocker):
```
Comment (nfx) [inline tests/unit/test_cli.py:10]: you can skip testing `cli.py`.

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/hive_metastore/tables.py (hunk 10 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/tables.py:282]: Fixed

File: src/databricks/labs/ucx/hive_metastore/tables.py (hunk 12 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/tables.py:251]: Sure

File: src/databricks/labs/ucx/hive_metastore/tables.py (hunk 8 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/tables.py:270]: Done

File: tests/unit/hive_metastore/test_tables.py (hunk 13 above)
Comment (FastLee) [reply tests/unit/hive_metastore/test_tables.py:119]: Done

File: tests/unit/hive_metastore/test_tables.py (hunk 15 above)
Comment (FastLee) [reply tests/unit/hive_metastore/test_tables.py:136]: Done

File: tests/unit/test_cli.py (hunk 16 above)
Comment (FastLee) [reply tests/unit/test_cli.py:10]: But why?

File: tests/unit/hive_metastore/test_tables.py (hunk 14 above)
Comment (FastLee) [reply tests/unit/hive_metastore/test_tables.py:127]: Done

File: src/databricks/labs/ucx/cli.py [hunk 17]
```diff
@@ -86,7 +113,7 @@ def main(raw):
         log_level = "info"
     databricks_logger = logging.getLogger("databricks")
     databricks_logger.setLevel(log_level.upper())
-
+    logger.info(payload)
```
Comment (nfx) [inline src/databricks/labs/ucx/cli.py:116]: Remove it from here, it's logged as debug in the cli already

File: src/databricks/labs/ucx/hive_metastore/mapping.py [hunk 18]
```diff
@@ -76,3 +82,27 @@ def load(self) -> list[Rule]:
...
+
+    def skip_table(self, backend: StatementExecutionBackend, schema: str, table: str):
+        # Marks a table to be skipped in the migration process
+        try:
+            backend.execute(f"ALTER TABLE `{schema}`.`{table}` SET TBLPROPERTIES('{self.UCX_SKIP_PROPERTY}' = true)")
+        except NotFound as nf:
+            if "[TABLE_OR_VIEW_NOT_FOUND]" in str(nf):
+                logger.error(f"Failed to apply skip marker for Table {schema}.{table}. Table not found.")
+            else:
+                logger.error(nf)
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/mapping.py:94]: Why do you need to catch exceptions just to log them?..

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/hive_metastore/mapping.py (hunk 18 above)
Comment (FastLee) [reply src/databricks/labs/ucx/hive_metastore/mapping.py:94]: To prevent the nasty log trace.

File: src/databricks/labs/ucx/cli.py (hunk 17 above)
Comment (nfx) [inline src/databricks/labs/ucx/cli.py:116]:
```suggestion
```

**Review by nfx — CHANGES_REQUESTED** [review]

**Review by nfx — APPROVED** [review]

**Pull request merged**